File: julia/ch1/Ex128.jl
"""
```jldoctest
julia> include("c:/Users/Yohan/IdeaProjects/sicp/julia/ch1/ExXXX.jl")

julia> ExXXX.func(arg)

or
shell> julia ExXXX.jl
```
"""
module Ex128

fast_times = 128

sq(n) = n * n

divides(a, b) = b % a == 0

function find_divisor(n, test)
    if sq(test) > n
        return n
    elseif divides(test, n)
        return test
    else
        return find_divisor(n, test + 1)
    end
end

smallest_divisor(n) = find_divisor(n, 2)

prime(n) = n == smallest_divisor(n)

function fermat_test(n)
    try_it(a) = expmod(a, n, n) == a
    return try_it(1 + random(n - 1))
end

function fast_prime(n, times)
    if times == 0
        return true
    elseif fermat_test(n)
        return fast_prime(n, times-1)
    else
        return false
    end
end 

function miller_rabin_test(n)
    try_it(a) = miller_rabin_expmod(a, n-1, n) == 1
    return try_it(1 + random(n - 1))
end

function fast_miller_rabin(n, times)
    if times == 0
        return true
    elseif miller_rabin_test(n)
        return fast_miller_rabin(n, times-1)
    else
        return false
    end
end 

function expmod(base, exp, m)
    if exp == 0
        return 1
    elseif divides(2, exp)
        return rem(sq(expmod(base, div(exp,2), m)), m)
    else
        return rem(base * expmod(base, exp-1, m), m)
    end
end

function non_trivial_square_root(x, square, m)
    result = (1 == square) && (x != 1) && (x != m-1)
    return result ? 0 : square
end

function miller_rabin_expmod(base, exp, m)
    squaremod_with_check(x) = non_trivial_square_root(x, rem(sq(x), m), m)
    if exp == 0
        return 1
    elseif divides(2, exp)
        return squaremod_with_check(miller_rabin_expmod(base, div(exp,2), m))
    else
        return rem(base * miller_rabin_expmod(base, exp-1, m), m)
    end
end

random(limit) = rem(abs(rand(Int, 1)[1]), limit)

function test_case(n)
    p = prime(n)
    fp = fast_prime(n, fast_times)
    mr = fast_miller_rabin(n, fast_times)
    return "$n \t Prime: $p \t Fast Prime: $fp \t Miller-Rabin: $mr"
end

end

# testing
# primes
println(Ex128.test_case(11))
println(Ex128.test_case(13))
println(Ex128.test_case(97))
# composite numbers
println(Ex128.test_case(10))
println(Ex128.test_case(77))
println(Ex128.test_case(6603))
# Carmichael numbers
println(Ex128.test_case(561))
println(Ex128.test_case(1105))
println(Ex128.test_case(1729))
println(Ex128.test_case(2465))
println(Ex128.test_case(2821))
println(Ex128.test_case(6601))

println("done")

File: ruby/ch1/ex128.rb
module Ex128

def divides(a, b)
  0 == b % a
end

def square(n)
  n * n
end

def find_divisor(n, test)
  if test * test > n
    n
  elsif divides(test, n)
    test
  else
    find_divisor(n, test + 1)
  end
end 

def smallest_divisor(n)
  find_divisor(n, 2)
end

def is_prime(n)
  n == smallest_divisor(n)
end

def fast_prime(n, times)
  if times == 0
    true
  elsif fermat_test(n)
    fast_prime(n, times-1)
  else
    false
  end
end

def fermat_test(n)
  try_it = lambda { |a| a == expmod(a, n, n) }
  try_it.call(1 + rand(n-1))
end

def expmod(base, exp, m)
  if exp == 0
    1
  elsif divides(2, exp)
    square(expmod(base, exp/2, m)) % m
  else
    (base * expmod(base, exp-1, m)) % m
  end
end

def fast_miller_rabin(n, times)
  if times == 0
    true
  elsif miller_rabin_test(n)
    fast_miller_rabin(n, times-1)
  else
    false
  end
end

def miller_rabin_test(n)
  try_it = lambda { |a| 1 == miller_rabin_expmod(a, n-1, n) }
  try_it.call(1 + rand(n-1))
end

def miller_rabin_expmod(base, exp, m)
  non_trivial_square_root = -> (x, sq) {
    test = (sq == 1) && (x != 1) && (x != (m-1))
    test ? 0 : sq
  }
  squaremod_with_check = -> (x) {
    non_trivial_square_root.call(x, square(x) % m)
  }
  if exp == 0
    1
  elsif divides(2, exp)
    squaremod_with_check.call(miller_rabin_expmod(base, exp/2, m))
  else
    (base * miller_rabin_expmod(base, exp-1, m)) % m
  end
end

def test_case(n)
  prime = is_prime(n)
  fp = fast_prime(n, 4)
  mr = fast_miller_rabin(n, 4)
  print("#{n} \t Prime: #{prime} \t Fast Prime:#{fp} \t Miller-Rabin:#{mr}\n")
end

end

include Ex128
print "Testing\n"

test_case(9)
test_case(10)
test_case(11)
test_case(71)
test_case(97)
test_case(561)
test_case(1105)
test_case(1729)
test_case(2465)
test_case(2821)
test_case(6601)

print "done", "\n"
